### ISSUE TITLE
Barriers?

### ISSUE THREAD

**Issue by krogovin** [gpuweb/gpuweb#862]
Looking at the WebGPU spec, I could not find any command for memory barriers, i.e. to specify that side effects on storage buffers from a previous draw of dispatch compute must be visible  in the next draw or dispatch compute within the same command buffer.

Or am I asking about this too soon?

**Comment by Degerz** [comment]
Do you mean the equivalent to Vulkan's vkMemoryBarrier and friends ? 

It has been already [decided](https://docs.google.com/document/d/1_2OZJx-83ho_rQ5V5QmKViqXhJulAXa150MxPNyirRQ/edit#) that memory barriers will be implicit just like on Metal ...

**Comment by kvark** [comment]
If you are asking about the barrier on the API side, they are implicit, and are actually [specified](https://gpuweb.github.io/gpuweb/#programming-model-synchronization).
TL;DR: there are no barriers within a pass (between draw calls), but there are barriers between compute dispatches.

**Comment by kainino0x** [comment]
To clarify:

> there are no **implicit\*** barriers within a pass (between draw calls), but there are **implicit\*** barriers between compute dispatches (and there are no explicit barriers (yet?)).

**Comment by krogovin** [comment]
I admit that making the memory barrier implicit like Metal is attractive for ease of use, but it leaves a fair amount of performance on the table as that it essentially forces a pipeline flush and associated cache flush for the buffer usage. I can see this having a significant negative performance impact on compute. Compounding it is that any browser implementation mapping to Vulkan or Direct3D 12 will have to issue the memory barriers for those API's even if the application does not need to. 

Given that part fo the spirit of WebGPU was to be explicit, this goes against that.

Can we change this question to issue? I think doing implicit memory barriers is not a good idea (even OpenGL required explicit barriers too).

**Comment by kainino0x** [comment]
Over a year of discussion, and significant hardware vendor input, went into the decision to have implicit memory barriers, and I don't think the decision is likely to change.

An important consideration here is that to ensure portability of applications, WebGPU has to verify that barriers are correct. Any rules that it would have for doing so would be no less strict than the implicit barriers that we have now - which means there's no opportunity for the application to omit barriers that it doesn't actually need. Instead, the app would need to express resource usages in such a way that those barriers are not required or inferred.

**Comment by krogovin** [comment]
> An important consideration here is that to ensure portability of applications, WebGPU has to verify that barriers are correct.

This claim is quite dubious. Indeed, an application inserting an extra barrier does harm correctness, so WebGPU would not need to verify that a barrier command is valid.  On the other hand, only an application's logic knows when the writes from a previous command are needed by a sub-sequent command. The lack of a barrier would just mean that the writes are not completed and it is an application logic error, not a portability error. Both Vulkan and Direct3D 12 are for a wide variety of hardware and they have barriers.

> Over a year of discussion, and significant hardware vendor input, went into the decision to have implicit memory barriers, and I don't think the decision is likely to change.

Is there a publicly (or privately) available link to these discussions? Being able to read them would shed light on the descision.

**Comment by kvark** [comment]
> Indeed, an application inserting an extra barrier does not harm correctness

That's right, but it's also useless technically

> so WebGPU would not need to verify that a barrier command is valid

That doesn't sound right. WebGPU would have to validated the barrier: it will need to see how it fits into all the transitions it knows about, whether it's compatible with regards to image layouts, access flags, etc, even if the barrier is doing extra unnecessary stalling.

> On the other hand, only an application's logic knows when the writes from a previous command are needed by a sub-sequent command

That question, I think is, cornerstone of our decision. If it's *only* the application knowing that, than we wouldn't know if the application code is correct in the first place. To ensure correctness and portability, *we* have to know whether anything from a previous command is needed by sub-sequent commands. So we have to do the tracking regardless, and we have to know which barriers are expected and where. Going from that to automatically inserting the barriers is a tiny step, and it makes developers lives much easier.

The only opportunity we missed is when perhaps the user can do a better job at placing barriers earlier, coalescing them, etc. We've yet to see how close the automatic barriers can get to optimal. And we have an option to extend the API in the future with something more explicit, if there is an opportunity there.

> I could not find any barrier stuff in GPUComputePassEncoder.

What I was saying is that we are automatically inserting barriers *inside* a compute pass between dispatches (unlike individual draws in a render pass). Users don't have get any barrier API today.

> Is there a publicly (or privately) available link to these discussions? Being able to read them would shed light on the descision.

That's true. A lot of the ones I remember were just verbal communications on GDC, various face-to-face conferences, etc.

**Comment by krogovin** [comment]
> The only opportunity we missed is when perhaps the user can do a better job at placing barriers earlier, coalescing them, etc. We've yet to see how close the automatic barriers can get to optimal. And we have an option to extend the API in the future with something more explicit, if there is an opportunity there.

Automatic barriers are from optimal for a very common scenario: multiple compute shaders dispatches write to the same buffer but at disjoint locations (or possible overlapping locations via atomic ops). The current interface for WebGPU will force a WebGPU implementation to issue a memory barrier between any two compute-dispatches that write to the same buffer (because of the timeline). All desktop GPU's (even the lowly Intel integrated GPU's) can have multiple shaders with also multiple binding tables active all at once. A memory barrier means basically a pipeline flush and a cache write flush. The larger the GPU (and the more incremental the nature of the successive dispatches) the worse the hit. The solutions for an application for this are ugly:

- An application could do an "uber" compute shader to make the multiple dispatches into a single one. This decreases performance of the compiled shader (by a quite significant margin sadly). In addition, various incremental algorithms will have a different natural way to break a compute job into work-items and work-groups.

- An application could instead use buffer views to declare that ranges of writes are disjoint. The really bad part is that outputted data from an incremental compute will then lose the ability for such output data to be interleaved which means when the compute shader that uses the data will suffer a great in cache efficiency. Worse some incremental compute jobs will write to the same memory locations on different passes via atomic opts which then makes this solution not possible.

Both of the above solutions means that an application is working around an API issue that should not exist. Given that two of the three targeted underneath API's have barriers (and in truth Metal does as well for bindless resources), the restriction is incorrect.

> To ensure correctness and portability, we have to know whether anything from a previous command is needed by subsequent commands.

This is not correct. The real issue is this: an application works fine on Metal backend but does not on a Vulkan or Direct3D 12 backends because the application did not issue the barriers. That is the crux. The point of an explicit interface is that it is an application's responsibility to do the right thing and be tested on multiple platforms.

I have several algorithms on my plate already that will suffer from significant inefficiencies with the two solutions outlined above to avoid a WebGPU backend from issuing automatic barriers.

**Comment by kvark** [comment]
> Automatic barriers are from optimal for a very common scenario: multiple compute shaders dispatches write to the same buffer but at disjoint locations (or possible overlapping locations via atomic ops)

This isn't a problem with automatic barriers. This is coming from the portability guarantees that the API sets up. Currently, it defines each dispatch to have its own synchronization scope, thus auto-inserting barriers between them. This matches Metal compute passes, and we haven't got a case where it do a lot of harm. If you have such an example, and we can craft a benchmark, that would be very helpful! I did argue for making the whole compute pass being a single sync scope in #59, and I'd still like us to have this option (perhaps, as a parameter on `GPUComputePassDescriptor`).

Worth noting that the API sees the ranges of buffers used by operations, so we can make it so there are no barriers between dispatches that use disjoint ranges, at least.

> An application could instead use buffer views to declare that ranges of writes are disjoint.

We don't have the API for buffer views. There were plans previously to get them, but currently there are none.

> This is not correct; no API can check if an application's logic is correct. The real issue is this: an application works fine on Metal backend but does not on a Vulkan or Direct3D 12 backends because the application did not issue the barriers. That is the crux. The point of an explicit interface is that it is an application's responsibility to do the right thing and be tested on multiple platforms.

We agree on the problem statement (users can misuse the barriers) but totally diverge on the solutions. I don't see it feasible to expect the users to test WebGPU applications on all platforms before shipping. We are making API in such a way that it's portable by specification, not by testing.

> I have several algorithms on my plate already that will suffer from significant inefficiencies with the two solutions outlined above to avoid a WebGPU backend from issuing automatic barriers.

Please do benchmark and share!

**Comment by Kangz** [comment]
Closing in favor of #59 that discusses adding some form manual control for barriers.
